numpy/core/src/multiarray/ctors.c

```diff
@@ -2269,21 +2273,28 @@ PyArray_FromArrayAttr(PyObject *op, PyArray_Descr *typecode, PyObject *context)
 NPY_NO_EXPORT PyArray_Descr *
 PyArray_DescrFromObject(PyObject *op, PyArray_Descr *mintype)
 {
-    PyArray_Descr *dtype;
+    PyArray_Descr *descr;
 
-    dtype = mintype;
-    Py_XINCREF(dtype);
+    int ndim;
+    npy_intp shape[NPY_MAXDIMS];
+    PyObject *out_dtypemeta;
+    coercion_cache_obj *coercion_cache = NULL;
 
-    if (PyArray_DTypeFromObject(op, NPY_MAXDIMS, &dtype) < 0) {
+    // TODO: In principle there is a tiny chance of behaviour change, due to
+    //       a different order of promotion mainly. This should be fine, but
+    //       possibly mentioned somewhere.
+    int res = PyArray_DiscoverDTypeAndShapeFromObject(
```

Comment:
Multiline comments still [should be formatted](https://numpy.org/devdocs/dev/style_guide.html) with `/* */` even though we now can use C99 comments for single line. Perhaps add the actual promotion case dtype?

Reply:
Yeah, I have a weird convention for myself currently that I use `//` for comments I think I should go back to and revise and `/*` style comments for normal comments.

Reply:
The whole recursion pattern in that function is currently messed up, probably first thing on my todo list to clean up.